src/receive_imf.rs

```diff
@@ -848,116 +1058,189 @@ pub async fn from_field_to_contact_id(
-    }
-
-    // check if the message introduces a new chat:
-    // - outgoing messages introduce a chat with the first to: address if they are sent by a messenger
-    // - incoming messages introduce a chat only for known contacts if they are sent by a messenger
-    // (of course, the user can add other chats manually later)
-    let to_id: ContactId;
-    let state: MessageState;
-    let hidden = is_reaction;
-    let mut needs_delete_job = false;
-    let mut restore_protection = false;
+        false
+    };
 
-    // if contact renaming is prevented (for mailinglists and bots),
-    // we use name from From:-header as override name
-    if prevent_rename {
-        if let Some(name) = &mime_parser.from.display_name {
-            for part in &mut mime_parser.parts {
-                part.param.set(Param::OverrideSenderDisplayname, name);
+    // Decide on the type of chat we assign the message to.
+    //
+    // The chat may not exist yet, i.e. there may be
+    // no database row and ChatId yet.
+    let mut num_recipients = mime_parser.recipients.len();
+    if from_id != ContactId::SELF {
+        let mut has_self_addr = false;
+        for recipient in &mime_parser.recipients {
+            if context.is_self_addr(&recipient.addr).await? {
+                has_self_addr = true;
             }
         }
+        if !has_self_addr {
+            num_recipients += 1;
+        }
     }
 
-    if chat_id.is_none() && is_mdn {
-        chat_id = Some(DC_CHAT_ID_TRASH);
-        info!(context, "Message is an MDN (TRASH).",);
-    }
+    let chat_assignment = if should_trash {
+        ChatAssignment::Trash
+    } else if let Some(grpid) = mime_parser.get_chat_group_id() {
+        if mime_parser.was_encrypted() {
+            ChatAssignment::GroupChat {
+                grpid: grpid.to_string(),
+            }
+        } else if let Some(parent) = &parent_message {
+            if let Some((chat_id, chat_id_blocked)) =
+                // Try to assign to a chat based on In-Reply-To/References.
+                lookup_chat_by_reply(context, mime_parser, parent, is_partial_download)
+                        .await?
+            {
+                // Try to assign to a chat based on In-Reply-To/References.
+                ChatAssignment::ExistingChat {
+                    chat_id,
+                    chat_id_blocked,
+                }
+            } else {
+                ChatAssignment::AdHocGroup
+            }
+        } else {
+            // Could be a message from old version
+            // with opportunistic encryption.
+            //
+            // We still want to assign this to a group
+            // even if it had only two members.
+            //
+            // Group ID is ignored, however.
+            ChatAssignment::AdHocGroup
+        }
+    } else if let Some(mailinglist_header) = mime_parser.get_mailinglist_header() {
+        let _listid = mailinglist_header_listid(mailinglist_header)?;
```

Comment:
Why this line? `mailinglist_header_listid()` will be called later, so, we will anyway return an error if it fails. Maybe it was left from an earlier iteration and can just be removed?

```suggestion
```